### PR TITLE
[MRG] Add noise attribute to skopt.GaussianProcessRegressor

### PR DESCRIPTION
**Pull request by MechCoder**
- [x] Adds a noise attribute directly to `skopt.learning.GaussianProcessRegressor`
- [x] At prediction time, to predict `P(f(x) | y)` the kernel `K(X_train, X_new)` and `K(X_new, X_new)`, should not include the noise or white kernel argument, this has been described in Eq, 2.24 of http://www.gaussianprocess.org/gpml/chapters/RW2.pdf . Note that `K(X_train, X_new)` and `K(X_new, X_new)` do not have the `sigma**2` term.

### CODE REVIEW

File: skopt/learning/gpr.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    """
+    GaussianProcessRegressor that allows noise tunability.
+    """
+
+    def fit(self, X, y):
+        super(GaussianProcessRegressor, self).fit(X, y)
+        for param, value in self.kernel_.get_params().items():
+            # XXX: Should return this only in the case where a
+            # WhiteKernel is added.
+            if param.endswith('noise_level'):
```
Comment (MechCoder) [inline skopt/learning/gpr.py]: This seems a bit hacky to me. What do you think are the other alternatives?

**Comment by MechCoder** [comment]
What are your thoughts about decoupling the noise argument from the `base_estimator` argument? This makes it explicit that I have added a `WhiteKernel` whose noise component I can set to zero while estimating `K(X_new, X_new)` and `K(X_train, X_new)`

**Comment by betatim** [comment]
Where would you want to introduce the `noise=` keyword? As an argument to `gp_minimize`? As a naive user I think I'd prefer that either I give a kernel that I've designed myself and it gets used or that `skopt` uses its default. Having a parameter that then (maybe) modifies the kernel that I passed seems like it would make things more complicated.

Not knowledgable enough about this whole magic of brewing a kernel :(

cc @glouppe

**Comment by glouppe** [comment]
or simply use the `alpha` parameter from the original scikit-learn interface? (instead of adding the `WhiteKernel` component)

**Comment by MechCoder** [comment]
> or simply use the alpha parameter from the original scikit-learn interface?

We can do that but that assumes I need to know the amount of noise before hand right?

The `alpha` parameter does the right thing, but using a `WhiteKernel` means I need to set the noise component of the kernel to zero while doing the kernel product here (https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/gaussian_process/gpr.py#L285) and here (https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/gaussian_process/gpr.py#L298). What are some other ways to achieve that?

**Comment by MechCoder** [comment]
Should be ready for reviews. I would consider this as a bug.

Before this PR (`n_calls=50`)

gp_before

After this PR (`n_calls=50`)

gp_after

File: skopt/learning/gpr.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        super(GaussianProcessRegressor, self).fit(X, y)
+        for param, value in self.kernel_.get_params().items():
+            if param.endswith('noise_level'):
+                self.noise_ = value
+                break
+
+        # The noise component of this kernel should be set to zero
+        # while estimating K(X, X_test) and K(X_test, X_test)
+        # Note that the term K(X, X) should include the noise but
+        # this (K(X, X))^-1y is precomputed as the attribute alpha.
```
Comment (glouppe) [inline skopt/learning/gpr.py]: What `alpha` attribute are you referring to?

Comment (glouppe) [inline skopt/learning/gpr.py]: Can you add the reference to Eqn. 2.24 in the comment?

**Comment by glouppe** [comment]
Also, not convinced by the before/after figure. Both look fine to me.

**Comment by MechCoder** [comment]
> but we should be aware this trick only holds for identically distributed noise, right?

Yeah, :/. I had added that point in the initial PR (https://github.com/scikit-optimize/scikit-optimize/blob/master/skopt/optimizer/gp.py#L69)

> Also, not convinced by the before/after figure. Both look fine to me.

Why do you think so, given a good number of function calls, the mean prediction should approximate the actual function right, and the first one seems a bit off?

File: skopt/learning/gpr.py (hunk 2 above)
Comment (MechCoder) [inline skopt/learning/gpr.py]: Sorry I meant this `alpha_` (https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/gaussian_process/gpr.py#L229)

**Comment by glouppe** [comment]
> Why do you think so, given a good number of function calls, the mean prediction should approximate the actual function right, and the first one seems a bit off?

This would be true for uniformly distributed samples, but this assumption does not hold in BO. We should only expect to be accurate in regions close of candidate optima. In the first figure, the region of high uncertainty is never ever sampled again because its LCB is larger than other parts of the input space. This is an expected behaviour of the algorithm.

**Comment by MechCoder** [comment]
Yes, that's true. Sorry about the "noise"

**Comment by MechCoder** [comment]
But I think the second cluster of points should be closer to the second local minima, but seems to be displaced by a bit in the first graph. But that just might be me nitpicking

File: skopt/learning/gpr.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
+from sklearn.gaussian_process import GaussianProcessRegressor as sk_GaussianProcessRegressor
+from sklearn.gaussian_process.kernels import WhiteKernel
+
```
Comment (MechCoder) [inline skopt/learning/gpr.py]: done

**Comment by MechCoder** [comment]
anything else?

**Comment by betatim** [comment]
I'm happy, but don't really understand much of the GP mumbojumbo so deferring to @glouppe 😃

File: skopt/learning/gpr.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        # Note that the term K(X, X) should include the noise but
+        # this (K(X, X))^-1y is precomputed as the attribute `alpha_`.
+        # (Notice the underscore).
+        # This has been described in Eq 2.24 of
+        # http://www.gaussianprocess.org/gpml/chapters/RW2.pdf
+        if isinstance(self.kernel_, WhiteKernel):
+            self.kernel_.set_params(noise_level=0.0)
+        else:
+            for param, value in self.kernel_.get_params().items():
+                if isinstance(value, WhiteKernel):
```
Comment (glouppe) [inline skopt/learning/gpr.py]: Isnt this only correct when `WhiteKernel` is a term in a sum kernel? (e.g. in the case where `kernel = K*WhiteKernel`, i dont believe this trick holds, but your if-statement will still be true)

Comment (MechCoder) [inline skopt/learning/gpr.py]: I never saw that operations like `+` inherited from `Sum`, so I pushed a workaround that and added a test

File: skopt/learning/gpr.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        # (Notice the underscore).
+        # This has been described in Eq 2.24 of
+        # http://www.gaussianprocess.org/gpml/chapters/RW2.pdf
+        if isinstance(self.kernel_, WhiteKernel):
+            self.kernel_.set_params(noise_level=0.0)
+        elif _check_WhiteKernel_in_Sum(self.kernel_):
+            for param, value in self.kernel_.get_params().items():
+                if isinstance(value, WhiteKernel):
+                    self.kernel_.set_params(
+                        **{param: WhiteKernel(noise_level=0.0)})
```
Comment (glouppe) [inline skopt/learning/gpr.py]: Sorry for nitpicking again, but this still does not solve completely the issue. If you have `kernel = K*WhiteKernel + WhiteKernel`, you would change both of them. 

The cleanest solution might be to instantiate the kernel yourself in this new class, adding in the `WhiteKernel` term to the user-provided kernel, and then keep a reference to this term.

Comment (MechCoder) [inline skopt/learning/gpr.py]: > adding in the WhiteKernel term to the user-provided kernel,

This goes against the principles of sklearn to not set logic at init time, right?

Comment (glouppe) [inline skopt/learning/gpr.py]: Not if you have something like:

```
def __init__(self, ..., kernel, ...):
    self.kernel = kernel

def fit(self, X, y):
    self.noise_ = WhiteKernel()
    self.kernel_ = self.kernel + self.noise_
    ...
    self.gp_ = sklearn.GaussianProcessRegressor(self.kernel_).fit(X, y)
    ...
    self.noise_.set_params(noise_level=0.0)
    ...
```

Comment (MechCoder) [inline skopt/learning/gpr.py]: But you wouldn't always want to set the "noise" option on, no? Which is why I suggested the `noise="auto"` option to suggest if we wanted to add the `WhiteKernel` or not.

Comment (MechCoder) [inline skopt/learning/gpr.py]: `noise="auto"` -> `noise="gaussian"` (to be more accurate)

Comment (MechCoder) [inline skopt/learning/gpr.py]: fixing this now

**Comment by MechCoder** [comment]
@glouppe Your solution will not work because the provided kernel is not modified at fit time, but rather a clone of it. I have pushed another fix (with tests). Let me know what you think.

File: skopt/learning/gpr.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+    alpha_: array-like, shape = (n_samples,)
+        Dual coefficients of training data points in kernel space
+
+    log_marginal_likelihood_value_: float
+        The log-marginal-likelihood of ``self.kernel_.theta``
+    """
+    def __init__(self, kernel=None, alpha=1e-10,
+                 optimizer="fmin_l_bfgs_b", n_restarts_optimizer=0,
+                 normalize_y=False, copy_X_train=True, random_state=None,
+                 noise=None):
```
Comment (glouppe) [inline skopt/learning/gpr.py]: So by default, `noise=None`, which amounts to reuse scikit-learn's GaussianProcessRegressor. Sorry to be blunt, but in the end, what is the added advantage of this PR? simply providing a shortcut for adding a `WhiteKernel` component? i.e. `noise="gaussian"` versus `kernel=K+WhiteKernel()`?

Comment (glouppe) [inline skopt/learning/gpr.py]: ... oh no, forget that comment. It also deals with disabling that `WhiteKernel` part when making predictions. Accordingly, it might be good to add a test highlighting the difference between `skopt.GP(kernel=K, noise="gaussian").predict` and `sklearn.GP(kernel=K+WhiteKernel()).predict`.

**Comment by glouppe** [comment]
LGTM!

**Comment by MechCoder** [comment]
@glouppe Anything else?

**Comment by MechCoder** [comment]
Phew, this took more time then I had expected.

**Comment by MechCoder** [comment]
squashed and mergred

**Comment by betatim** [comment]
:shipit: good job!
